airbyte-integrations/connectors/source-snowflake/source_snowflake/authenticator.py

```diff
@@ -0,0 +1,90 @@
+from cryptography.hazmat.backends import default_backend
+
+
+class SnowflakeJwtAuthenticator(JwtAuthenticator):
+    TOKEN_DURATION = 55
+    SNOWFLAKE_URL_SUFFIX = ".snowflakecomputing.com"
+    HTTP_PREFIX = "https://"
+
+    def get_auth_header(self) -> Mapping[str, Any]:
+        """
+        :return: A dictionary containing all the necessary headers to authenticate.
+        """
+        return {
+            'Authorization': f"Bearer {self.token}"
+        }
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+
+    @classmethod
+    def from_config(cls, config):
+        sub = cls.get_formatted_account(config['host'], config["credentials"]['user_name'])
+        formatted_private_key = cls._format_private_key(config["credentials"]['private_key'])
+        public_key_finger_print = cls.calculate_public_key_fingerprint(formatted_private_key, sub)
+        kwargs = {"config": {},
+                  "algorithm": JwtAlgorithm.RS256,
+                  "iss": public_key_finger_print,
+                  "sub": sub,
+                  "secret_key": formatted_private_key,
+                  'parameters': {},
+                  'token_duration': cls.TOKEN_DURATION}
+        return cls(**kwargs)
+
+    @classmethod
+    def get_formatted_account(cls, account: str, user_name: str) -> str:
+        """
+        Format the account and user_name to satisfy JWT token formatting
+        """
+        processed_account = account
+        if account.endswith(cls.SNOWFLAKE_URL_SUFFIX):
+            processed_account = processed_account.split(cls.SNOWFLAKE_URL_SUFFIX)[0]
+
+        if account.startswith(cls.HTTP_PREFIX):
+            processed_account = processed_account.split(cls.HTTP_PREFIX)[1]
+
+        processed_account = processed_account.split('.')[0].upper()
+        return f"{processed_account}.{user_name.upper()}"
+
+    @classmethod
+    def calculate_public_key_fingerprint(cls, private_key: Text, sub: Text) -> Text:
+        """
+        Given a private key in string format, return the public key fingerprint.
+        :param private_key: private key string
+        :return: public key fingerprint
+        """
+        private_key_object = load_pem_private_key(private_key.encode(),
+                                           None,
+                                           default_backend())
+
+        # Get the raw bytes of public key.
+        public_key_raw = private_key_object.public_key().public_bytes(Encoding.DER, PublicFormat.SubjectPublicKeyInfo)
+
+        # Get the sha256 hash of the raw bytes.
+        sha256hash = hashlib.sha256()
+        sha256hash.update(public_key_raw)
+
+        # Base64-encode the value and prepend the prefix 'SHA256:'.
+        public_key_fp = f"{sub}.SHA256:{base64.b64encode(sha256hash.digest()).decode('utf-8')}"
+
+        return public_key_fp
+
+    @classmethod
+    def _format_private_key(cls, private_key: Text):
+        start_private_key = "-----BEGIN PRIVATE KEY-----"
```

Comment:
Following the steps [here](https://docs.snowflake.com/en/user-guide/key-pair-auth), this gives me an error. I think it is because I created an encrypted key (which seems to be the safer approach) and the headers/footer are `-----BEGIN ENCRYPTED PRIVATE KEY-----` and `-----END ENCRYPTED PRIVATE KEY-----`. Just changing the code for that does not work because we don't pass the password in `load_pem_private_key`

Can you provide the step by step of how you set up the credentials? On my side, I also tried:
* Create the private key: `openssl genrsa 2048 | openssl pkcs8 -topk8 -inform PEM -out rsa_key.p8 -nocrypt`
* Create the public key: `openssl rsa -in rsa_key.p8 -pubout -out rsa_key.pub`
* Run `ALTER USER AIRBYTEDEVELOPERTESTACCOUNT SET RSA_PUBLIC_KEY=<public key with the break lines>` and get `Statement executed successfully.`
* Copy/paste private key in secrets

Expected result: I can run a `discover` command using
Actual result: `401 Client Error: Unauthorized for url: https://<obfuscated url>.snowflakecomputing.com/api/v2/statements?requestId=ef514c8f-da57-49f1-afc1-2c311394609a&async=false` with log `"JWT token is invalid. [a49f6aa6-9bea-42f5-af46-f5fd105e86a9]` which seems to come from the HTTP response

Reply:
I answered that in the video and I will write down the steps in docs/integrations/sources/snowflake.md